src/coreclr/tools/Common/TypeSystem/Ecma/EcmaModule.cs

```diff
@@ -186,6 +186,11 @@ internal EcmaModule(TypeSystemContext context, PEReader peReader, MetadataReader
             _moduleResolver = customModuleResolver != null ? customModuleResolver : context;
         }
 
+        public void CleanResolvedTokens()
+        {
+            _resolvedTokens = new EcmaObjectLookupHashtable(this);
```

Comment:
This effectively drops the whole type system except for the `EcmaModule`. Would it be better to just drop `EcmaModule` as well instead of cleaning it up like this?

Reply:
`EcmaModule` contains `_peReader` field which was created during file opening. So `EcmaModule` dropping lead to reopen of all files on each loop iteration. But the optimization concludes to move file opening work `CompilerTypeSystemContext.AddModule()` out from loop.

Reply:
Can we keep the `_peReader` around and create a fresh EcmaModule from it instead of flushing it like this?

Also, would it be profitable to keep the full type system around for different compilations and only flush it e.g. when the memory goes over certain treashold?

Reply:
> Can we keep the `_peReader` around and create a fresh EcmaModule from it instead of flushing it like this?

I think it is possible, but it looks to be more complicated. What is the reasons to create whole fresh `EcmaModule` with old fields instead of just cleaning this hash table?
 
> Also, would it be profitable to keep the full type system around for different compilations and only flush it e.g. when the memory goes over certain treashold?

For answer to this performance graph (optimized/noinputbubble) may be useful. It seems that `CleanResolvedTokens` was included in performance graph as CPU_TIME in `Program.Run`, it is 1.21ms, <0.01%. Besides `CleanResolvedTokens` it is neccesary to restart `SetSystemModule` - 56.67ms, 0.36%. So keep the full _typeSystemContext variable could reduce less onle than 0.37%.

Reply:
> What is the reasons to create whole fresh EcmaModule with old fields instead of just cleaning this hash table?

Cleaner architecture.

Reply:
We should add APIs on the CompilerTypeSystemContext so that we can populate the hash table of open modules with the already created `PEReader`s.

Then when we need to reset the compiler context, we would just `new ReadyToRunCompilerContext(...)` instead of calling `Reconstruct` and give it the `PEReader`s we already have open. (It could e.g. be a new constructor in `ReadyToRunCompilerContext` that takes a `ReadyToRunCompilerContext` as a parameter so that it can access the internal state we want to copy).

This `CleanResolvedTokens` and the equivalent `Reconstruct` APIs on ReadyToRunCompilerContext will not work. `ReadyToRunCompilerContext` has a bunch of state in its base types we would also need to reset and that just leads to unsustainable architecture and a bug tail in a codepath we're not testing (I was initially only looking at the changes in `src/coreclr/tools/Common/` but I already see that more changes would be needed there to support `Reconstruct` and it's not clean changes).